core/src/test/java/org/apache/calcite/test/RelMetadataTest.java

```diff
@@ -971,6 +989,1025 @@ final RelMetadataFixture sql(String sql) {
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            ImmutableBitSet.of(), bitSetOf(2)));
+    Set<RelOptForeignKey> confirmedForeignKeys =
+        Sets.newHashSet(
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMPNULLABLES_QNAME, 7, true),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, true))),
+            bitSetOf(1), bitSetOf(2)),
+        RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMPNULLABLES_QNAME, 7, true),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, true))),
+            bitSetOf(3), bitSetOf(2)));
+    Set<RelOptForeignKey> foreignKeysIgnoreNull =
+        Sets.newHashSet(
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMPNULLABLES_QNAME, 7, false),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            bitSetOf(1), ImmutableBitSet.of()),
+        RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMPNULLABLES_QNAME, 7, false),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            bitSetOf(3), ImmutableBitSet.of()));
+    foreignKeysIgnoreNull.addAll(confirmedForeignKeys);
+    foreignKeysIgnoreNull.addAll(foreignKeys);
+    sql("select dept.name, empnullables.deptno, dept.deptno, empnullables.deptno "
+        + "from dept, empnullables")
+        .assertForeignKeys(
+            equalTo(foreignKeys), equalTo(EMPTY_FOREIGN_KEY_SET), false)
+        .assertForeignKeys(
+            equalTo(foreignKeysIgnoreNull), equalTo(confirmedForeignKeys), true);
+  }
+
+  @Test void testForeignKeysLeftOuterJoin() {
+    Set<RelOptForeignKey> foreignKeys =
+        Sets.newHashSet(
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            ImmutableBitSet.of(), bitSetOf(3)),
+        RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMP_QNAME, 7, false),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, false))),
+            bitSetOf(1), ImmutableBitSet.of()));
+    Set<RelOptForeignKey> confirmedForeignKeys =
+        Sets.newHashSet(
+            RelOptForeignKey.of(
+            Lists.newArrayList(
+                Pair.of(
+                    InferredConstraintKey.of(EMP_QNAME, 7, true),
+                    InferredConstraintKey.of(DEPT_QNAME, 0, true))),
+            bitSetOf(1), bitSetOf(3)));
```

Comment:
These two variables are used in almost every single test I see. They have too much duplicate code. You can define two methods that take different parameter types and then construct those two variables. This reduces code duplication.